tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/jms/multiprotocol/JMSMessageExpiryTest.java

```diff
@@ -0,0 +1,197 @@
+   }
+
+   @Test
+   @Timeout(30)
+   public void testOpenWireMaxExpiryDelayWithExpiration() throws Exception {
+      testExpiry(createOpenWireConnection(), DelayType.MAX, true);
+   }
+
+   @Test
+   @Timeout(30)
+   public void testCoreMinExpiryDelayWithExpiration() throws Exception {
+      testExpiry(createCoreConnection(), DelayType.MIN, true);
+   }
+
+   @Test
+   @Timeout(30)
+   public void testAmqpMinExpiryDelayWithExpiration() throws Exception {
+      testExpiry(createConnection(), DelayType.MIN, true);
+   }
+
+   @Test
+   @Timeout(30)
+   public void testOpenWireMinExpiryDelayWithExpiration() throws Exception {
+      testExpiry(createOpenWireConnection(), DelayType.MIN, true);
+   }
+
+   @Test
+   @Timeout(30)
+   public void testCoreMessageNeverExpire() throws Exception {
+      testExpiry(createCoreConnection(), DelayType.NEVER, true);
+   }
+
+   @Test
+   @Timeout(30)
+   public void testAmqpMessageNeverExpire() throws Exception {
+      testExpiry(createConnection(), DelayType.NEVER, true);
+   }
+
+   @Test
+   @Timeout(30)
+   public void testOpenWireMessageNeverExpire() throws Exception {
+      testExpiry(createOpenWireConnection(), DelayType.NEVER, true);
+   }
+
+   private void testExpiry(Connection connection, DelayType delayType, boolean setTimeToLive) throws Exception {
+      AddressSettings addressSettings = new AddressSettings();
+      if (delayType == DelayType.NORMAL) {
+         addressSettings.setExpiryDelay(EXPIRY_DELAY);
+      } else if (delayType == DelayType.MIN) {
+         addressSettings.setMinExpiryDelay(EXPIRY_DELAY);
+      } else if (delayType == DelayType.MAX) {
+         addressSettings.setMaxExpiryDelay(EXPIRY_DELAY);
+      } else if (delayType == DelayType.NEVER) {
+         addressSettings.setNeverExpire(true);
+      }
+      server.getAddressSettingsRepository().addMatch(getQueueName(), addressSettings);
+
+      Session s = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+      Queue q = s.createQueue(getQueueName());
+      MessageProducer p = s.createProducer(q);
+      if (setTimeToLive) {
+         if (delayType == DelayType.MIN) {
+            p.setTimeToLive(EXPIRY_DELAY / 2);
+         } else if (delayType == DelayType.MAX) {
+            p.setTimeToLive(EXPIRY_DELAY * 2);
+         } else if (delayType == DelayType.NEVER) {
+            p.setTimeToLive(EXPIRY_DELAY);
+         }
+      }
+      long start = System.currentTimeMillis();
+      p.send(s.createMessage());
+      p.close();
+      MessageConsumer c = s.createConsumer(q);
+      connection.start();
+      Message m = c.receive(100);
+      assertNotNull(m);
+      if (delayType == DelayType.NEVER) {
+         assertEquals(0, m.getJMSExpiration());
+      } else {
+         long duration = System.currentTimeMillis() - start;
+         long delayOnMessage = m.getJMSExpiration() - System.currentTimeMillis();
+         assertTrue(delayOnMessage >= (EXPIRY_DELAY - duration - WINDOW));
```

Comment:
Feels like WINDOW shouldnt really be doing anything (at least after using same reference for both calcs above) since you are already deducting a 'duration' which is including the consumer receive (and an assertNotNull) which necessarily makes it >= the broker-side elapsed time which could have influenced the actual expiration stamped by the broker, i.e you should already be deducting more than enough with just that?

Reply:
I recall that being necessary for tests to pass at one point. I guess I changed something along the way that eliminated the need for it.